File: B-Trucks/userSettings.swift
//
//  userSettings.swift
//  B-Trucks
//
//  Created by Edward M. Dempsey Jr. on 4/5/17.
//  Copyright © 2017 Wesley Van As. All rights reserved.
//

import UIKit
import Foundation

class userSettings: UIViewController {
    
    @IBOutlet var FirstName: UITextField!
    @IBOutlet var LastName: UITextField!
    @IBOutlet var Email: UITextField!
    @IBOutlet var Phone: UITextField!
    @IBOutlet var Password: UITextField!
    @IBOutlet var Password2: UITextField!
    
    
    let userID = UserDefaults.standard.object(forKey: "userID") as! String
    
    
    struct GlobalVariable {
        static var msg = String()
        static var pWord = String()
        static var pWord2 = String()
    }
    
    
    @IBAction func submit(_ sender: AnyObject) {
        
        
        
        print("part 2")
        
        func validateName(candidate: String) -> Bool {
            let nameRegex = "[A-Z][A-Za-z -]+"
            return NSPredicate(format: "SELF MATCHES %@", nameRegex).evaluate(with: candidate)
        }
        
        var fNameValidated = validateName(candidate: FirstName.text!)
        
        // allows for empty string to be posted
        if FirstName.text! == "" {
            fNameValidated = true
            print("part 3")
        }
        
        //print(fNameValidated)
        if fNameValidated == false {
            let alert = UIAlertController(title: "Invalid First Name",
                                          message: "Please enter a valid first name.",
                                          preferredStyle: .alert)
            let cancelAction = UIAlertAction(title: "OK",
                                             style: .cancel, handler: nil)
            alert.addAction(cancelAction)
            self.present(alert, animated: true)
        }
        
        var lNameValidated = validateName(candidate: LastName.text!)
        
        // allows for empty string to be posted
        if LastName.text! == "" {
            lNameValidated = true
        }
        
        if lNameValidated == false {
            let alert = UIAlertController(title: "Invalid Last Name",
                                          message: "Please enter a valid last name.",
                                          preferredStyle: .alert)
            let cancelAction = UIAlertAction(title: "OK",
                                             style: .cancel, handler: nil)
            alert.addAction(cancelAction)
            self.present(alert, animated: true)
        }
        
        
        func validateEmail(candidate: String) -> Bool {
            let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}"
            return NSPredicate(format: "SELF MATCHES %@", emailRegex).evaluate(with: candidate)
        }
        var emailValidated = validateEmail(candidate: Email.text!)
        
        // allows for empty string to be posted
        if Email.text! == "" {
            emailValidated = true
        }
        //print(emailValidated)
        
        if emailValidated == false {
            let alert = UIAlertController(title: "Invalid Email",
                                          message: "Please enter a valid email address.",
                                          preferredStyle: .alert)
            let cancelAction = UIAlertAction(title: "OK",
                                             style: .cancel, handler: nil)
            alert.addAction(cancelAction)
            self.present(alert, animated: true)
        }
        
        func validatePassword(candidate: String) -> Bool {
            let passwordRegex = "(?=.*[a-z])(?=.*[A-Z])(?=.*\\d).{6,15}"
            return NSPredicate(format: "SELF MATCHES %@", passwordRegex).evaluate(with: candidate)
        }
        
        var validatedPassword = validatePassword(candidate: Password.text!)
    
        // allows for empty string to be posted
        if Password.text! == "" && Password2.text! == "" {
            validatedPassword = true
            GlobalVariable.pWord = "x"
            GlobalVariable.pWord2 = "x"
            print("part 5")
        } else {
            GlobalVariable.pWord = Password.text!
            GlobalVariable.pWord2 = Password2.text!
        }
        
        if validatedPassword == false {
            let alert = UIAlertController(title: "Invalid Password",
                                          message: "Please enter a password containing at least one capital, one lowercase, one number and is between 8 and 14 characters.",
                                          preferredStyle: .alert)
            let cancelAction = UIAlertAction(title: "OK",
                                             style: .cancel, handler: nil)
            alert.addAction(cancelAction)
            self.present(alert, animated: true)
        }
        
        
                if validatedPassword && GlobalVariable.pWord == GlobalVariable.pWord2 {
                    var passwordsMatch = true
                }
                else {
                    let alert = UIAlertController(title: "Passwords do not match",
                                                  message: "Please make sure that both passwords match",
                                                  preferredStyle: .alert)
                    let cancelAction = UIAlertAction(title: "OK",
                                                     style: .cancel, handler: nil)
                    alert.addAction(cancelAction)
                    self.present(alert, animated: true)
                }
        
        func validatePhone(candidate: String) -> Bool {
            let phoneRegex = "\\d{10}"
            return NSPredicate(format: "SELF MATCHES %@", phoneRegex).evaluate(with: candidate)
        }
        
        var phoneValidated = validatePhone(candidate: Phone.text!)
        
        // allows for empty string to be posted
        if Phone.text! == "" {
            phoneValidated = true
        }
        
        if phoneValidated == false {
            let alert = UIAlertController(title: "Invalid Phone",
                                          message: "Please enter a valid phone number with the following format: 8888888888",
                                          preferredStyle: .alert)
            let cancelAction = UIAlertAction(title: "OK",
                                             style: .cancel, handler: nil)
            alert.addAction(cancelAction)
            self.present(alert, animated: true)
        }
        
        
        print("part 6")
        if emailValidated && fNameValidated && lNameValidated && phoneValidated {
            
            let myUrl = URL(string: "http://cgi.soic.indiana.edu/~team18/userSettings.php")
            
            var request = URLRequest(url: myUrl!)
            request.httpMethod = "POST"
            let postString = "fName=\(FirstName.text!)&lName=\(LastName.text!)&email=\(Email.text!)&phone=\(Phone.text!)&password=\(GlobalVariable.pWord)&userid=\(userID)"
            
            print(postString)
            print("part 7")
            
            request.httpBody = postString.data(using: String.Encoding.utf8)
            
            let task = URLSession.shared.dataTask(with: request as URLRequest) {
                data, response, error in
                
                print("part 4")
                
                if error != nil {
                    print("error=\(error)")
                    return
                }
                
                // creating a string
                
                do {
                    //converting response to NSDictionary
                    let myJSON = try JSONSerialization.jsonObject(with: data!, options: .mutableContainers) as? NSDictionary
                    print("part 8")
                    //parsing the json
                    if let parseJSON = myJSON {
                        //getting the json response
                        GlobalVariable.msg = (parseJSON["message"] as! String?)!
                        print("part 9")
                        if GlobalVariable.msg != "Changes successfully made!" {
                            //Sends an alert to the main block's que to notify the user that the email they registered with already exists in the database.
                            
                            OperationQueue.main.addOperation {
                                let alert = UIAlertController(title: "Changes Unsuccessful",
                                                              message: "Changes were not successfully saved. Please try again.",
                                                              preferredStyle: .alert)
                                let cancelAction = UIAlertAction(title: "OK",
                                                                 style: .cancel, handler: nil)
                                alert.addAction(cancelAction)
                                self.present(alert, animated: true)
                            }
                        } else {
                            //sends an alert to the main block's que to notify the user that they have successfully registered.
                            OperationQueue.main.addOperation {
                                let alert = UIAlertController(title: "Changes Successfully Made!",
                                                              message: "Your changes will be reflected in our system immediately",
                                                              preferredStyle: .alert)
                                //Sends the user back to the login screen when they click OK on the alert.
                                let OKAction = UIAlertAction(title: "OK",
                                                             style: UIAlertActionStyle.default, handler: {
                                                                (_)in
                                                                self.performSegue(withIdentifier: "unwindToLogin", sender: self)
                                })
                                alert.addAction(OKAction)
                                self.present(alert, animated: true)
                            }
                            
                        }
                        
                        //print("response = \(type(of:GlobalVariable.msg))")
                        //print(GlobalVariable.msg)
                        //print(GlobalVariable.existingEmail)
                    }
                } catch {
                    print(error)
                }
            }
            task.resume()
        }
    }
}

File: userSettings.php
<?php
	$host = 'db.soic.indiana.edu';
	$user = 'caps16_team18';
	$password = 'my+sql=btrucks2017';
	$database = 'caps16_team18';

	$conn = mysqli_connect($host, $user, $password, $database);

	$response = array();


	$firstName = $_POST['fName'];
	$lastName = $_POST['lName'];
	$email = $_POST['email'];
	$phone = $_POST['phone'];
	$password = $_POST['password'];
	$uid = $_POST['userid'];

	if(!$conn){
	    die('Connection Failed: ' . mysqli_connect_error());
	  } else {
	  			
	  			//creating statement
	  			if($firstName != "" && $lastName != "" && $email != "" && $phone != "" && $password != "x") { 
	  				
	  				$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."', Email = '".$email."', Phone = '".$phone."', Password = '".$password."' WHERE UserID = '".$uid."';";
	  				
	  				$result = mysqli_query($conn, $query);
	  			
	  			} elseif($firstName != "" && $lastName != "" && $email != "" && $phone != "") { 
	  				
	  				$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."', Email = '".$email."', Phone = '".$phone."' WHERE UserID = '".$uid."';";
	  				
	  				$result = mysqli_query($conn, $query);

	  			} elseif($firstName != "" && $lastName != "" && $phone != "" && $password != "x") { 
	  				
	  				$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."', Phone = '".$phone."', Password = '".$password."' WHERE UserID = '".$uid."';";
	  				
	  				$result = mysqli_query($conn, $query);

	  			} elseif($firstName != "" && $phone != "" && $email != "" && $password != "x") { 
	  				
	  				$query = "UPDATE Users SET Fname = '".$firstName."', Phone = '".$phone."', Email = '".$email."', Password = '".$password."' WHERE UserID = '".$uid."';";
	  				
	  				$result = mysqli_query($conn, $query);

	  			} elseif($phone != "" && $lastName != "" && $email != "" && $password != "x") { 
	  				
	  				$query = "UPDATE Users SET Phone = '".$phone."', Lname = '".$lastName."', Email = '".$email."', Password = '".$password."' WHERE UserID = '".$uid."';";
	  				
	  				$result = mysqli_query($conn, $query);

	  			} elseif($firstName != "" && $lastName != "" && $email != "" && $password != "x") { 
	  				
	  				$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."', Email = '".$email."', Password = '".$password."' WHERE UserID = '".$uid."';";
	  				
	  				$result = mysqli_query($conn, $query);

	  			} elseif($firstName && $lastName && $email) {
  					$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."', Email = '".$email."' WHERE UserID = '".$uid."';";
  				
  	
  					$result = mysqli_query($conn, $query);

	  			} elseif($firstName && $lastName && $password != "x") {
  						$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."', Password = '".$password."' WHERE UserID = '".$uid."';";
  				
  	
  						$result = mysqli_query($conn, $query);
				} elseif($firstName && $email && $password != "x") {
					$query = "UPDATE Users SET Fname = '".$firstName."', Email = '".$email."', Password = '".$password."' WHERE UserID = '".$uid."';";
		

					$result = mysqli_query($conn, $query);

				} elseif($lastName && $email && $password != "x") {
					$query = "UPDATE Users SET Lname = '".$lastName."', Email = '".$email."', Password = '".$password."' WHERE UserID = '".$uid."';";
	

					$result = mysqli_query($conn, $query);

				} elseif($firstName && $lastName && $phone) {
  					$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."', Phone = '".$phone."' WHERE UserID = '".$uid."';";
  				
  	
  					$result = mysqli_query($conn, $query);

  				} elseif($firstName && $phone && $email) {
  					$query = "UPDATE Users SET Fname = '".$firstName."', Phone = '".$phone."', Email = '".$email."' WHERE UserID = '".$uid."';";
  				
  	
  					$result = mysqli_query($conn, $query);

  				} elseif($phone && $lastName && $email) {
  					$query = "UPDATE Users SET Phone = '".$phone."', Lname = '".$lastName."', Email = '".$email."' WHERE UserID = '".$uid."';";
  				
  	
  					$result = mysqli_query($conn, $query);

  				} elseif($firstName && $password != "x" && $phone) {
  					$query = "UPDATE Users SET Fname = '".$firstName."', Password = '".$password."', Phone = '".$phone."' WHERE UserID = '".$uid."';";
  				
  	
  					$result = mysqli_query($conn, $query);

  				} elseif($phone && $lastName && $password != "x") {
  					$query = "UPDATE Users SET Phone = '".$Phone."', Lname = '".$lastName."', Password = '".$password."' WHERE UserID = '".$uid."';";
  				
  	
  					$result = mysqli_query($conn, $query);

  				} elseif($phone && $password != "x" && $email) {
  					$query = "UPDATE Users SET Phone = '".$phone."', Password = '".$password."', Email = '".$email."' WHERE UserID = '".$uid."';";
  				
  	
  					$result = mysqli_query($conn, $query);

				} elseif($firstName && $lastName) {
					$query = "UPDATE Users SET Fname = '".$firstName."', Lname = '".$lastName."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($firstName && $email) {
					$query = "UPDATE Users SET Fname = '".$firstName."', Email = '".$email."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($firstName && $password != "x") {
					$query = "UPDATE Users SET Fname = '".$firstName."', Password = '".$password."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($lastName && $email) {
					$query = "UPDATE Users SET Lname = '".$lastName."', Email = '".$email."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($lastName && $password != "x") {
					$query = "UPDATE Users SET Lname = '".$lastName."', Password = '".$password."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($email && $password != "x") {
					$query = "UPDATE Users SET Email = '".$email."', Password = '".$password."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($firstName && $phone) {
					$query = "UPDATE Users SET Fname = '".$firstName."', Phone = '".$phone."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($lastName && $phone) {
					$query = "UPDATE Users SET Lname = '".$lastName."', Phone = '".$phone."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($email && $phone) {
					$query = "UPDATE Users SET Email = '".$email."', Phone = '".$phone."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($phone && $password != "x") {
					$query = "UPDATE Users SET Phone = '".$phone."', Password = '".$password."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($firstName) {
					$query = "UPDATE Users SET Fname = '".$firstName."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($lastName) {
					$query = "UPDATE Users SET Lname = '".$lastName."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($email) {
					$query = "UPDATE Users SET Email = '".$email."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($phone) {
					$query = "UPDATE Users SET Phone = '".$phone."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

				} elseif($password != "x") {
					$query = "UPDATE Users SET Password = '".$password."' WHERE UserID = '".$uid."';";


					$result = mysqli_query($conn, $query);

	 				} else {
	 					echo "didn't work";

	   			}
	  			}

			if ($result) {
				// return 0 means changes successful
				$response['error'] = false;
				$response['message'] = "Changes successfully made!";
				
			} else {
				// return 1 means failed to make changes
				$response['error'] = true;
				$response['message'] = mysqli_error();
			}
		 
		
		echo json_encode($response);
	 



	mysqli_close($conn);	
?>
